src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -77,17 +109,186 @@ public String getScript() {
 
     /**
      * Method for getting the type.
-     * @return The type of script -- dynamic, indexed, or file.
+     * 
+     * @return The type of script -- inline, indexed, or file.
      */
     public ScriptType getType() {
         return type;
     }
 
     /**
+     * Method for getting language.
+     * 
+     * @return The language of the script to be compiled/executed.
+     */
+    public String getLang() {
+        return lang;
+    }
+
+    /**
      * Method for getting the parameters.
+     * 
      * @return The map of parameters the script will be executed with.
      */
     public Map<String, Object> getParams() {
         return params;
     }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        script = in.readString();
+        type = ScriptType.readFrom(in);
+        lang = in.readOptionalString();
+        if (in.readBoolean()) {
+            params = in.readMap();
+        }
+        doReadFrom(in);
+    }
+
+    protected void doReadFrom(StreamInput in) throws IOException {
+        // For sub-classes to Override
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(script);
+        ScriptType.writeTo(type, out);
+        out.writeOptionalString(lang);
+        boolean hasParams = params != null;
+        out.writeBoolean(hasParams);
+        if (hasParams) {
+            out.writeMap(params);
+        }
+        doWriteTo(out);
+    }
+
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        // For sub-classes to Override
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params builderParams) throws IOException {
+        // NOCOMMIT decide if we want to output the shorthand just string (not object) version. Has implications for parsing, specifically 
+        // with multiple source transforms. Run TransformOnIndexMapperIntegrationTest.getTransformed() with seed -Dtests.seed=CCF6041A004DDD9D to see why
```

Comment:
@uboness I have made the type `@Nullable` internally to the class so that the `Script(String script)` constructor can set the type as `null` but the `Script(String script, ScriptType type, String lang, Map<String, Object> params)` constructor still errors if not type is passed in.